workflow-manager/task/task.go

```diff
@@ -0,0 +1,253 @@
+	// until completion functions passed to any and all calls to Enqueue() have
+	// returned.
+	Enqueue(task Task, completion func(error))
+	// Stop blocks until all tasks passed to Enqueue() have been enqueued in the
+	// underlying system, and all completion functions pased to Enqueue() have
+	// returned, and so it is safe to exit the program without losing any tasks.
+	Stop()
+}
+
+// CreatePubSubTopic creates a PubSub topic with the provided ID, as well as a
+// subscription with the same ID that can later be used by a facilitator.
+// Returns error on failure.
+func CreatePubSubTopic(project string, topicID string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return fmt.Errorf("pubsub.newClient: %w", err)
+	}
+
+	topic, err := client.CreateTopic(ctx, topicID)
+	if err != nil {
+		return fmt.Errorf("pubsub.CreateTopic: %w", err)
+	}
+
+	tenMinutes, _ := time.ParseDuration("10m")
+
+	subscriptionConfig := pubsub.SubscriptionConfig{
+		Topic:            topic,
+		AckDeadline:      tenMinutes,
+		ExpirationPolicy: time.Duration(0), // never expire
+	}
+	if _, err := client.CreateSubscription(ctx, topicID, subscriptionConfig); err != nil {
+		return fmt.Errorf("pubsub.CreateSubscription: %w", err)
+	}
+
+	return nil
+}
+
+// GCPPubSubEnqueuer implements Enqueuer using GCP PubSub
+type GCPPubSubEnqueuer struct {
+	topic     *pubsub.Topic
+	waitGroup sync.WaitGroup
+	dryRun    bool
+}
+
+// NewGCPPubSubEnqueuer creates a task enqueuer for a given project and topic
+// in GCP PubSub. If dryRun is true, no tasks will actually be enqueued. Clients
+// should re-use a single instance as much as possible to enable batching of
+// publish requests.
+func NewGCPPubSubEnqueuer(project string, topicID string, dryRun bool) (*GCPPubSubEnqueuer, error) {
+	client, err := pubsub.NewClient(context.Background(), project)
+	if err != nil {
+		return nil, fmt.Errorf("pubsub.NewClient: %w", err)
+	}
+
+	return &GCPPubSubEnqueuer{
+		topic:  client.Topic(topicID),
+		dryRun: dryRun,
+	}, nil
+}
+
+func (e *GCPPubSubEnqueuer) Enqueue(task Task, completion func(error)) {
+	e.waitGroup.Add(1)
+	go func(task Task) {
+		defer e.waitGroup.Done()
+		jsonTask, err := json.Marshal(task)
+		if err != nil {
+			completion(fmt.Errorf("marshaling task to JSON: %w", err))
+			return
+		}
+
+		if e.dryRun {
+			log.Printf("dry run, not enqueuing task")
+			completion(nil)
+			return
+		}
+
+		// Publish() returns immediately, giving us a handle to the result that we
+		// can block on to see if publishing succeeded. The PubSub client
+		// automatically retries for us, so we just keep the handle so the caller
+		// can do whatever they need to after successful publication and we can
+		// block in Stop() until all tasks have been enqueued
+		res := e.topic.Publish(context.Background(), &pubsub.Message{Data: jsonTask})
```

Comment:
technically possible but really would encourage to use a timing out ctx here.
```
ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
``` 

Plus you want to group the two usages of context (Publish + Get)